src/ray/common/runtime_env_manager.cc

```diff
@@ -0,0 +1,28 @@
+#include "ray/common/runtime_env_manager.h"
+#include "ray/util/logging.h"
+namespace ray {
+
+void RuntimeEnvManagerBase::AddUriReference(const std::string &hex_id,
+                                            const rpc::RuntimeEnv &runtime_env) {
+  if (!runtime_env.working_dir_uri().empty()) {
+    const auto &uri = runtime_env.working_dir_uri();
+    uri_reference_[uri]++;
+    id_to_uris_[hex_id].push_back(uri);
+  }
+}
+
+void RuntimeEnvManagerBase::RemoveUriReference(const std::string &hex_id) {
+  for (const auto &uri : id_to_uris_[hex_id]) {
+    --uri_reference_[uri];
+    auto ref_cnt = uri_reference_[uri];
+    RAY_CHECK(ref_cnt >= 0);
+    if (ref_cnt == 0) {
+      uri_reference_.erase(uri);
+      RAY_LOG(INFO) << "Deleting uri: " << uri;
+      DeleteURI(uri);
+    }
+  }
```

Comment:
Could there be a race condition where a reference is added at the same time we are trying to GC the zip?

Reply:
Please correct me if I'm wrong.

io_service is running in a single thread, so no race condition in RuntimeEnvManagerBase itself.

There are two functions: `HandleJobStarted` and `HandleJobFinished`, which can't run together (from my understanding, io_service can only schedule one function a time).

If `HandleJobStarted` run first, it'll lock the resource, and HandleJobFinished will not delete the local resource.
If `HandleJobFinished` run first, it'll delete the local resource, and later `HandleJobStarted` will run. After `HandleJobStarted` locks the resource, the worker will check whether it exists locally or not and fetch it when it's missing (https://github.com/ray-project/ray/blob/ad8e35b9195a84722ea0accf2a63785bd39911c4/python/ray/worker.py#L1239-L1246).


But I think there is one race condition which is about uploading. If two drivers are uploading the same file, it'll have a race condition. To fix this, we should fix it at the DB level with txn.

Reply:
Hmmmmm, one other thought and I think I'm wrong here. Here `RuntimeEnvCleanup` actually is running in another process, so it might be the cases, `HandleJobStarted` and `RuntimeEnvCleanup` can run at the same time. One easy fixing could be to make it a sync call, which is similar like we are calling a python funciton.


Or we have a deleting status introduced and wait there if there is any deleting pending. This one gives me the feeling that it's complicated, but still ok.

Reply:
Hmm a sync call would stall the event loop. I think the case we care most about is if deletion is running at the same time as setup, in this case the new job would not be setup correctly.

This is making me think we should be moving the Creation method into an IO worker as well, so that we can manage both in a consistent way and avoid race conditions. So the RuntimeEnvWorker would be used for both Creation and Cleanup.

> If two drivers are uploading the same file, it'll have a race condition. 

This is fine since they would be uploading the same contents right?